lib/ecto/adapters/postgres/connection.ex

```diff
@@ -820,7 +820,14 @@ if Code.ensure_loaded?(Postgrex) do
     defp default_type(list, {:array, inner} = type) when is_list(list) do
       ["ARRAY[",  Enum.map(list, &default_type(&1, inner)) |> Enum.intersperse(?,), "]::", ecto_to_db(type)]
     end
-    defp default_type(literal, _type) when is_binary(literal),  do: [?', escape_string(literal), ?']
+    defp default_type(literal, _type) when is_binary(literal) do
+      if :binary.match(literal, <<0>>) == :nomatch && String.valid?(literal) do
+        [?', escape_string(literal), ?']
+      else
+        raise(ArgumentError, "default values are interpolated as UTF-8 strings, and cannot contain null bytes. `#{inspect literal}` is invalid. " <>
```

Comment:
I like this a lot and I have one suggestion: what if we include the encoded string in the error message? So something like this:

```elixir
"default values are interpolated as UTF-8 strings and cannot contain null bytes. " <>
  "`#{inspect literal}` is invalid. If you want to write it as a binary, use #{inspect(encoded)}, " <>
  "otherwise refer to PostgreSQL documentation for instructions on how to escape this SQL type"
```

Reply:
That's a good point. Given that we clearly say that "if you want to have it as a binary", we should be covered. Plus, if you send a poorly utf-8 encoded binary to a text column, I am hoping postgres will provide a better error message (error than a protocol error).